### PR TITLE
Program subscriptions

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
It would be useful to be able to track account changes by program owner, but there is currently no easy way to do that.

#### Summary of Changes
Add program subscriptions to rpc_pubsub

I would especially appreciate eyes on the accounts apis I added for this, as there may be better ways to do this.

### CODE REVIEW

**Comment by jackcmay** [comment]
Getting changes during the registration is sufficient for watching what is currently happening but it implicitly includes the requirement that the listener be listening while the all changes it cares about takes place.  The token exchange is one example where the client may come and go, be started out of order, etc.., but still needs to know the current state of the accounts at the time of registration in order to make sense of any changes to accounts after that.  Is there a way to get the current state at the time of registration?  I guess it could request each account individually but that seems like a lot of back and forth and not atomic...  Thoughts?

This is a larger issue that applies to this pr but also the other subscriptions.

File: book/src/jsonrpc-api.md [hunk 1]
```diff
@@ -296,6 +298,53 @@ Unsubscribe from account userdata change notifications
 
 ---
 
+### programSubscribe
+Subscribe to a program to receive notifications when the userdata for a given account owned by the program changes
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply book/src/jsonrpc-api.md]: Correct!

File: core/src/rpc_subscriptions.rs [hunk 2]
```diff
@@ -139,6 +170,19 @@ impl RpcSubscriptions {
             }
         }
 
+        let programs: Vec<_> = {
+            let subs = self.program_subscriptions.read().unwrap();
+            subs.keys().cloned().collect()
+        };
+        for program_id in &programs {
+            let accounts = &bank.get_program_accounts_modified_since_parent(program_id);
+            if !accounts.is_empty() {
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply core/src/rpc_subscriptions.rs]: Nope and yes!

**Comment by CriesofCarrots** [comment]
@jackcmay , once these Accounts methods exist, it should be relatively easy to add an RPC to get the current state of program-owned accounts.
Does that sound sufficient?
(The other subscriptions already have matching RPC methods for current state.)
We may need to consider truncating/paginating results from such a call... your use case will probably show us if that's necessary ;)

**Comment by mvines** [comment]
Pagination will definitely be needed if we build a way to fetch all accounts for a program id.  But how about no?  Let’s explore alternatives first, such as linking accounts to each other or using something off-chain to track them.  Let’s discuss in discord or on a call sometime

File: runtime/src/accounts.rs [hunk 3]
```diff
@@ -1805,4 +1837,36 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_load_by_program() {
+        let paths = get_tmp_accounts_path!();
+        let accounts_db = AccountsDB::new(0, &paths.paths);
+
+        // Load accounts owned by various programs into AccountsDB
+        let pubkey0 = Keypair::new().pubkey();
+        let account0 = Account::new(1, 0, Pubkey::new(&[2; 32]));
```
Comment (mvines) [inline runtime/src/accounts.rs]: account0 and account1 have the same pubkey, that’s ok?

Comment (CriesofCarrots) [reply runtime/src/accounts.rs]: They are specifically owned by the same program id, yeah. To test whether the methods actually return multiples.

**Comment by CriesofCarrots** [comment]
> Looks good to me, are you doing to do the web3.js plumbing too?

Sure, I'll take a crack at it.

**Pull request merged**